dbms/src/Flash/Mpp/MPPTunnel.h

```diff
@@ -203,23 +210,99 @@ class AsyncTunnelSender : public TunnelSender
     GRPCSendQueue<TrackedMppDataPacketPtr> queue;
 };
 
-/// LocalTunnelSender just provide readForLocal method to return one element one time
-/// LocalTunnelSender is owned by the associated ExchangeReceiver
+template <bool enable_fine_grained_shuffle>
 class LocalTunnelSender : public TunnelSender
 {
 public:
-    using Base = TunnelSender;
-    using Base::Base;
-    TrackedMppDataPacketPtr readForLocal();
+    LocalTunnelSender(
+        size_t source_index_,
+        const String & req_info_,
+        ExchangeReceiverBase * recv_base_,
+        const LoggerPtr & log_,
+        size_t queue_size,
+        MemoryTrackerPtr & memory_tracker_,
+        const String & tunnel_id_)
+        : TunnelSender(queue_size, memory_tracker_, log_, tunnel_id_)
+        , source_index(source_index_)
+        , req_info(req_info_)
+        , recv_base(recv_base_)
+        , is_done(false)
+    {}
+
+    ~LocalTunnelSender() override
+    {
+        closeLocalTunnel(false, "");
+    }
+
+    bool push(TrackedMppDataPacketPtr && data) override
+    {
+        auto myid = std::this_thread::get_id();
+        std::stringstream ss;
+        ss << myid;
+        std::string tid = ss.str();
+
+        if (unlikely(checkPacketErr(data)))
+            return false;
+
+        // receiver_mem_tracker pointer will always be valid because ExchangeReceiverBase won't be destructed
+        // before all local tunnels are destructed so that the MPPTask which contains ExchangeReceiverBase and
+        // is responsible for deleting receiver_mem_tracker must be destroyed after these local tunnels.
+        MemoryTracker * receiver_mem_tracker = recv_base->getMemoryTracker();
+        data->switchMemTracker(receiver_mem_tracker);
+
+        auto res = pushPacket<enable_fine_grained_shuffle, false, true>(source_index, req_info, data, recv_base->getMsgChannels(), log);
```

Comment:
Can you wrap a new class `ChannelWriter`?
```
class ChannelWriter
{
    std::vector<MsgChannelPtr> * msg_channels;
    bool enable_fine_grained_shuffle;

    void write(size_t source_index) { xxx }
}
```
And it can used by `LocalTunnel`, `AsyncRequestHandler` and `ExchangeReceiver::readLoop`

Reply:
> Can you wrap a new class `ChannelWriter`?
> 
> ```
> class ChannelWriter
> {
>     std::vector<MsgChannelPtr> * msg_channels;
>     bool enable_fine_grained_shuffle;
> 
>     void write(size_t source_index) { xxx }
> }
> ```
> 
> And it can used by `LocalTunnel`, `AsyncRequestHandler` and `ExchangeReceiver::readLoop`

Done